File: src/com/jetbrains/idear/Idear.java
package com.jetbrains.idear;

import com.intellij.ide.plugins.IdeaPluginDescriptor;
import com.intellij.ide.plugins.PluginManager;
import com.intellij.openapi.components.ApplicationComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.extensions.PluginId;
import org.jetbrains.annotations.NotNull;

public class Idear implements ApplicationComponent {


    public Idear() {
    }

    @Override
    public void initComponent() {
        ServiceManager.getService(ASRService.class).init();
        ServiceManager.getService(GrammarService.class).init();
        initTTSService();
    }

    private void initTTSService() {
        PluginId id = PluginId.getId("com.jetbrains.idear");
        IdeaPluginDescriptor plugin = PluginManager.getPlugin(id);
        assert plugin != null;

        ClassLoader current = Thread.currentThread().getContextClassLoader();
        try {
            ClassLoader classLoader = plugin.getPluginClassLoader();
            Thread.currentThread().setContextClassLoader(classLoader);
            //nasty hack to load TTSService with appropriate constructor
            TTSService service = ServiceManager.getService(TTSService.class);
        } finally {
            Thread.currentThread().setContextClassLoader(current);
        }
    }

    @Override
    public void disposeComponent() {
        ServiceManager.getService(ASRService.class).dispose();
        ServiceManager.getService(TTSService.class).dispose();
    }

    @Override
    @NotNull
    public String getComponentName() {
        return "Idear";
    }
}

File: src/main/java/org/openasr/idear/Idear.kt
package org.openasr.idear

import com.intellij.ide.plugins.PluginManager
import com.intellij.openapi.Disposable
import com.intellij.openapi.extensions.PluginId
import org.openasr.idear.asr.ASRService
import org.openasr.idear.tts.TTSService

object Idear : Disposable {
    // TODO: PluginManager.getPlugin is deprecated, but this alternative is only supported in #193 (2019.3)
    //       Currently we're supporting <idea-version since-build="131"/> which is pre-2016
//    val plug = PluginManagerCore.getPlugin(PluginId.getId("com.jetbrains.idear"))
    @Suppress("DEPRECATION")
    val plugin = PluginManager.getPlugin(PluginId.getId("com.jetbrains.idear"))!!

    override fun dispose() {
        ASRService.dispose()
        TTSService.dispose()
    }
}